### PR TITLE
Implement backup purge command to purge orphaned snapshots

### PR DESCRIPTION
**Pull request by dkropachev**
Closes https://github.com/scylladb/scylla-manager/issues/3441

There could be the case when backup task is deleted, but snapshots are left on the location.
So, we need to have a command that deletes these stale snapshots

### CODE REVIEW

File: pkg/service/backup/service.go [hunk 1]
```diff
@@ -1187,3 +1175,133 @@ func (s *Service) DeleteSnapshot(ctx context.Context, clusterID uuid.UUID, locat
...
+	if tags.IsEmpty() {
+		return nil, nil
+	}
+
+	nodeIDManifests := groupManifestsByNode(manifests)
+	var mut sync.Mutex
+
+	var runFunc parallel.RunMapElem[string, Manifests]
+
+	runFunc = func(nodeID string, m Manifests) error {
```
Comment (Michal-Leszczynski) [inline pkg/service/backup/service.go:1247]: Why not just:
```go
runFunc := func(nodeID string, m Manifests) error {
```
Or do you think that it is more descriptive this way?

File: pkg/restapi/backup_test.go [hunk 2]
```diff
@@ -191,8 +309,11 @@ func TestBackupListFiles(t *testing.T) {
 	cm.EXPECT().GetCluster(gomock.Any(), cluster.ID.String()).Return(cluster, nil)
 	bm.EXPECT().ListFiles(gomock.Any(), cluster.ID, locations, filter).Return(golden, nil)
 
-	r := withForm(listBackupFilesRequest(cluster.ID), locations, filter, cluster.ID.String())
+	r := withForm(listBackupFilesRequest(cluster.ID), false, locations, filter, cluster.ID.String())
 	w := httptest.NewRecorder()
 	h.ServeHTTP(w, r)
-	assertJsonBody(t, w, golden)
+	err := assertJsonBody(w, golden)
+	if err != nil {
```
Comment (Michal-Leszczynski) [inline pkg/restapi/backup_test.go:316]: Maybe combine those two into a single line?

File: pkg/restapi/backup_test.go [hunk 3]
```diff
@@ -142,10 +257,13 @@ func TestBackupListAllClusters(t *testing.T) {
 	cm.EXPECT().GetCluster(gomock.Any(), cluster.ID.String()).Return(cluster, nil)
 	bm.EXPECT().List(gomock.Any(), cluster.ID, locations, filter).Return(golden, nil)
 
-	r := withForm(listBackupsRequest(cluster.ID), locations, filter, "")
+	r := withForm(listBackupsRequest(cluster.ID), false, locations, filter, "")
 	w := httptest.NewRecorder()
 	h.ServeHTTP(w, r)
-	assertJsonBody(t, w, golden)
+	err := assertJsonBody(w, golden)
+	if err != nil {
```
Comment (Michal-Leszczynski) [inline pkg/restapi/backup_test.go:264]: Maybe combine those two into a single line?

File: pkg/service/backup/list.go [hunk 4]
```diff
@@ -22,13 +22,13 @@ func listManifestsInAllLocations(ctx context.Context, client *scyllaclient.Clien
...
-		if _, ok := locations[hi.Location]; ok {
+	for idx := range hosts {
+		if _, ok := locations[hosts[idx].Location]; ok {
 			continue
 		}
-		locations[hi.Location] = struct{}{}
+		locations[hosts[idx].Location] = struct{}{}
 
-		lm, err := listManifests(ctx, client, hi.IP, hi.Location, clusterID)
+		lm, err := listManifests(ctx, client, hosts[idx].IP, hosts[idx].Location, clusterID)
```
Comment (Michal-Leszczynski) [inline pkg/service/backup/list.go:31]: What is the motivation behind this change?

File: pkg/service/backup/parallel.go [hunk 5]
```diff
@@ -32,8 +32,8 @@ func makeHostsLimit(hosts []hostInfo, limits []DCLimit) map[string]hostsLimit {
 	}
 
 	m := make(map[string]hostsLimit, len(dcLimit)+1)
-	for _, h := range hosts {
-		dc := h.DC
+	for idx := range hosts {
+		dc := hosts[idx].DC
```
Comment (Michal-Leszczynski) [inline pkg/service/backup/parallel.go:36]: What is the motivation behind this change?

File: pkg/service/backup/validation.go [hunk 6]
```diff
@@ -176,9 +176,9 @@ func (s *Service) Validate(ctx context.Context, clusterID, taskID, runID uuid.UU
 		p := newPurger(client, h.IP, log.NopLogger)
 
 		hostForNodeID := func() string {
-			for _, h := range hosts {
-				if h.ID == nodeID {
-					return h.IP
+			for idx := range hosts {
+				if hosts[idx].ID == nodeID {
+					return hosts[idx].IP
```
Comment (Michal-Leszczynski) [inline pkg/service/backup/validation.go:181]: What is the motivation behind this change?

File: pkg/service/backup/worker_schema.go [hunk 7]
```diff
@@ -99,12 +99,12 @@ func (w *worker) UploadSchema(ctx context.Context, hosts []hostInfo) (stepError
...
-	for _, hi := range hosts {
-		locations[hi.Location.String()] = hi
+	for idx := range hosts {
+		locations[hosts[idx].Location.String()] = hosts[idx]
 	}
 	hostPerLocation := make([]hostInfo, 0, len(locations))
-	for _, hi := range locations {
-		hostPerLocation = append(hostPerLocation, hi)
+	for key := range locations {
+		hostPerLocation = append(hostPerLocation, locations[key])
```
Comment (Michal-Leszczynski) [inline pkg/service/backup/worker_schema.go:107]: What is the motivation behind this change?

**Review by Michal-Leszczynski — CHANGES_REQUESTED** [review]
Sorry for large amount of comments, most of them are some nit picks that can probably be ignored. I'm curious why there were so many changes in terms of iterating over value vs index in this PR.

**Comment by dkropachev** [comment]
@karol-kokoszka , not yet, I will complete it this week

**Pull request closed without merging**